File: haskell/src/DataStructures/Deque.hs
{-|
Module      : DataStructures.Deque
Description : Double-ended queue implementation
Stability   : stable

A purely functional deque (double-ended queue).
Time: O(1) amortized for all operations
Space: O(n)
-}
module DataStructures.Deque
  ( Deque
  , empty
  , isEmpty
  , pushFront
  , pushBack
  , popFront
  , popBack
  , peekFront
  , peekBack
  , size
  , fromList
  , toList
  ) where

-- | A deque using two lists for amortized O(1) operations at both ends
data Deque a = Deque [a] [a] Int
  deriving (Show, Eq)

-- | Create an empty deque
empty :: Deque a
empty = Deque [] [] 0

-- | Check if deque is empty
isEmpty :: Deque a -> Bool
isEmpty (Deque _ _ n) = n == 0

-- | Add element to the front
pushFront :: a -> Deque a -> Deque a
pushFront x (Deque front back n) = Deque (x : front) back (n + 1)

-- | Add element to the back
pushBack :: a -> Deque a -> Deque a
pushBack x (Deque front back n) = Deque front (x : back) (n + 1)

-- | Remove element from the front
popFront :: Deque a -> Maybe (a, Deque a)
popFront (Deque [] [] _) = Nothing
popFront (Deque [] back n) = popFront (Deque (reverse back) [] n)
popFront (Deque (x:xs) back n) = Just (x, Deque xs back (n - 1))

-- | Remove element from the back
popBack :: Deque a -> Maybe (a, Deque a)
popBack (Deque [] [] _) = Nothing
popBack (Deque front [] n) = popBack (Deque [] (reverse front) n)
popBack (Deque front (x:xs) n) = Just (x, Deque front xs (n - 1))

-- | Peek at front element without removing
peekFront :: Deque a -> Maybe a
peekFront (Deque [] [] _) = Nothing
peekFront (Deque [] back _) = Just (Prelude.last back)
peekFront (Deque (x:_) _ _) = Just x

-- | Peek at back element without removing
peekBack :: Deque a -> Maybe a
peekBack (Deque [] [] _) = Nothing
peekBack (Deque front [] _) = Just (Prelude.last front)
peekBack (Deque _ (x:_) _) = Just x

-- | Get the number of elements
size :: Deque a -> Int
size (Deque _ _ n) = n

-- | Create from list (first element at front)
fromList :: [a] -> Deque a
fromList xs = Deque xs [] (length xs)

-- | Convert to list (front element first)
toList :: Deque a -> [a]
toList (Deque front back _) = front ++ reverse back

File: go/datastructures/deque.go
package datastructures

// Deque implements a double-ended queue using a doubly linked list.
// Time: O(1) for all operations
// Space: O(n)
type Deque[T any] struct {
	head *dequeNode[T]
	tail *dequeNode[T]
	size int
}

type dequeNode[T any] struct {
	value T
	prev  *dequeNode[T]
	next  *dequeNode[T]
}

func NewDeque[T any]() *Deque[T] {
	return &Deque[T]{}
}

func (d *Deque[T]) PushFront(value T) {
	node := &dequeNode[T]{value: value, next: d.head}
	if d.head != nil {
		d.head.prev = node
	} else {
		d.tail = node
	}
	d.head = node
	d.size++
}

func (d *Deque[T]) PushBack(value T) {
	node := &dequeNode[T]{value: value, prev: d.tail}
	if d.tail != nil {
		d.tail.next = node
	} else {
		d.head = node
	}
	d.tail = node
	d.size++
}

func (d *Deque[T]) PopFront() (T, bool) {
	var zero T
	if d.head == nil {
		return zero, false
	}
	value := d.head.value
	d.head = d.head.next
	if d.head != nil {
		d.head.prev = nil
	} else {
		d.tail = nil
	}
	d.size--
	return value, true
}

func (d *Deque[T]) PopBack() (T, bool) {
	var zero T
	if d.tail == nil {
		return zero, false
	}
	value := d.tail.value
	d.tail = d.tail.prev
	if d.tail != nil {
		d.tail.next = nil
	} else {
		d.head = nil
	}
	d.size--
	return value, true
}

func (d *Deque[T]) PeekFront() (T, bool) {
	var zero T
	if d.head == nil {
		return zero, false
	}
	return d.head.value, true
}

func (d *Deque[T]) PeekBack() (T, bool) {
	var zero T
	if d.tail == nil {
		return zero, false
	}
	return d.tail.value, true
}

func (d *Deque[T]) IsEmpty() bool {
	return d.size == 0
}

func (d *Deque[T]) Size() int {
	return d.size
}

func (d *Deque[T]) Clear() {
	d.head = nil
	d.tail = nil
	d.size = 0
}
